sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/ShareDirectoryStorageResourceContainer.cs

```diff
@@ -176,5 +176,59 @@ await ShareDirectoryClient.CreateIfNotExistsAsync(
                 options: options,
                 cancellationToken: cancellationToken).ConfigureAwait(false);
         }
+
+        protected override async Task ValidateTransferAsync(StorageResource sourceResource, CancellationToken cancellationToken = default)
+        {
+            CancellationHelper.ThrowIfCancellationRequested(cancellationToken);
+
+            if (sourceResource is ShareDirectoryStorageResourceContainer sourceShareDirectory)
+            {
+                // Make sure the transfer is supported (NFS -> NFS and SMB -> SMB)
+                if ((ResourceOptions?.IsNfs ?? false) != (sourceShareDirectory.ResourceOptions?.IsNfs ?? false))
+                {
+                    throw Errors.ShareTransferNotSupported();
+                }
+
+                // validate the source protocol
+                if (!sourceShareDirectory.ResourceOptions?.SkipProtocolValidation ?? true)
+                {
+                    try
+                    {
+                        ShareClient sourceParentShare = sourceShareDirectory.ShareDirectoryClient.GetParentShareClient();
+                        ShareProperties sourceProperties = await sourceParentShare.GetPropertiesAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+                        ShareProtocols setProtocolSource = (sourceShareDirectory.ResourceOptions?.IsNfs ?? false) ? ShareProtocols.Nfs : ShareProtocols.Smb;
+                        ShareProtocols actualProtocolSource = sourceProperties.Protocols ?? ShareProtocols.Smb;
+                        if (actualProtocolSource != setProtocolSource)
+                        {
+                            throw Errors.ProtocolSetMismatch("source", setProtocolSource, actualProtocolSource);
+                        }
+                    }
+                    catch (RequestFailedException ex) when (ex.Status == 403)
+                    {
+                        throw Errors.NoShareLevelPermissions("source");
+                    }
```

Comment:
If we do keep source/dest separate, maybe move all this to a helper function since its repeated below?